internal/pkg/siderolink/logstorage/logstorage.go

```diff
@@ -0,0 +1,302 @@
+// Copyright (c) 2024 Sidero Labs, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+
+// Package logstorage implements storage for storing logs of machines on the filesystem.
+package logstorage
+
+import (
+	"bytes"
+	"crypto/sha256"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"sync"
+
+	"github.com/hashicorp/go-multierror"
+	"github.com/klauspost/compress/zstd"
+	"github.com/siderolabs/gen/optional"
+	"go.uber.org/zap"
+)
+
+// CompressionExtension is the extension used for compressed log files.
+const CompressionExtension = "zst"
+
+// Storage stores logs for machines on file system.
+type Storage struct {
+	dirtyMachineIDs map[string]struct{}
+	logger          *zap.Logger
+	Path            string
+	Compress        bool
+	lock            sync.Mutex
+}
+
+// New creates a new Storage.
+func New(path string, compress bool, logger *zap.Logger) *Storage {
+	return &Storage{
+		Path:            path,
+		Compress:        compress,
+		dirtyMachineIDs: map[string]struct{}{},
+		logger:          logger,
+	}
+}
+
+// MarkDirty marks the machine with the given ID as dirty.
+func (l *Storage) MarkDirty(id string) {
+	l.lock.Lock()
+	defer l.lock.Unlock()
+	l.dirtyMachineIDs[id] = struct{}{}
+}
+
+// Exists returns true if the log file exists for the given machine ID.
+func (l *Storage) Exists(id string) (bool, error) {
+	_, err := os.Stat(l.logPath(id, l.Compress))
+	if err == nil {
+		return true, nil
+	}
+
+	if !errors.Is(err, os.ErrNotExist) {
+		return false, err
+	}
+
+	_, err = os.Stat(l.logPath(id, !l.Compress))
+	if err == nil {
+		return true, nil
+	}
+
+	if !errors.Is(err, os.ErrNotExist) {
+		return false, err
+	}
+
+	return false, nil
+}
+
+// Load loads logs of the machine with the given id, if exists, into the given writer.
+func (l *Storage) Load(id string, writer io.Writer) error {
+	compressed := l.Compress
+
+	bufferDataOptional, hashHexActual, err := l.readLogFile(id, l.Compress)
```

Comment:
Can we detect file compression by the file extension instead of trying to read it twice?

Reply:
ah, so the ID doesn't have the extension and we probe the files here

Reply:
Though I think something like this might make the code more readable:

```go
compressed, err := l.probeLogFile(id)
if err != nil {
  return err
}

bufferDataOptional, hashHexActual, err := l.readLogFile(id, compressed)
```